### PR TITLE
REGR: Preserve order by default in Index.intersection

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/24959

### CODE REVIEW

**Comment by TomAugspurger** [comment]
If we do this, we still need to decide the long-term behavior.

I thought that we eventually wanted to make `Index.union()` preserve the original order if possible (`sort=False`).

**Comment by TomAugspurger** [comment]
What exactly are you -1 on?

File: pandas/core/indexes/datetimes.py [hunk 1]
```diff
@@ -594,14 +594,22 @@ def _wrap_setop_result(self, other, result):
...
+    def intersection(self, other, sort=False):
         """
         Specialized intersection for DatetimeIndex objects. May be much faster
         than Index.intersection
 
         Parameters
         ----------
         other : DatetimeIndex or array-like
+        sort : bool, default True
+            Sort the resulting MultiIndex if possible
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]:
```suggestion
            Sort the resulting index if possible
```

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
> What exactly are you -1 on?

The change is done and its consistent with other set ops, which sort.  Sure it broke, but reverting makes this more confusing.

**Comment by jreback** [comment]
> In #24967 (comment) you said you want long-term sort=False. Is that correct or a typo?

yes, but for ALL set ops. Having one different is pretty odd.

**Comment by TomAugspurger** [comment]
Maybe? But having this one be different is clearly preferable to breaking from the previous behavior, right?

**Comment by jreback** [comment]
> Maybe? But having this one be different is clearly preferable to breaking from the previous behavior, right?

no I don't think so. Its unfortunate that we didn't have any testing around this, but I don't see a reason to create confusion . what's done is done.

**Comment by TomAugspurger** [comment]
K. We'll need input from others then; I think getting us back to the pre-0.24.0 behavior is a good idea.

I think @jorisvandenbossche also prefers changing it back in 0.24.1.

**Comment by samspills** [comment]
Since needing input from others is explicitly mentioned I'm dropping a comment even though I've never contributed to pandas. The default `sort=True` option broke some unit tests for me this morning, and is what led me to this issue. 

I would expect the default behaviour to be `sort=False` because:
1. creating/reading a dataframe or resetting the index doesn't sort automatically, so I would expect to receive an unsorted index when operating on it
2. returning just the index series does not return a sorted index so I would expect to receive an unsorted index when operating on it

Also, I think reasonable default behaviour moving forward is more important than any confusion that might arise from 0.24.0 vs. 0.24.1. 

Thanks for all the work on pandas, we rely on it extensively. Hope my comments are not intrusive

**Comment by TomAugspurger** [comment]
Thanks @samspills.

> Also, I think reasonable default behaviour moving forward is more important than any confusion that might arise from 0.24.0 vs. 0.24.1.

Just to be clear, the long-term goal is to have `sort=False` for all of our index setops. The question is how do we get there from 0.24.0.

1. Implement a deprecation warning for the current behavior (sort=True), and eventually change the default to `sort=False`.
2. Do a quick "break" from 0.24.0 to restore the 0.23.4 and earlier behavior with `sort=False` as the default. No deprecation cycle needed.

IMO, 2 is easier on users and downstream libraries. For people who never want intersection to sort, it's easiest to require `pandas != 0.24.0`, which has only been out for a few days.

If we go with option 1, all downstream users wanting to avoid sorting and support multiple versions of pandas will need to update all their calls to Index.intersection.

**Comment by jorisvandenbossche** [comment]
I also just commented about it on the issue: https://github.com/pandas-dev/pandas/issues/24959#issuecomment-458214689 Shall we continue the discussion there or here?

@samspills thanks for chiming in!

**Comment by jorisvandenbossche** [comment]
If we want `sort=False` as the long term behaviour for all set ops, then I agree with Tom: let's just quickly revert `Index.intersection` to have `sort=False` as default. That will make the life of people that depend on it easier.

**Comment by samspills** [comment]
From a user perspective, I think it would be preferable to blacklist a specific pandas version in our setup.py than to require either `<0.24` or `>=0.24.0`+specify sort behavour explicitly 😄

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
> 1. Implement a deprecation warning for the current behavior (sort=True), and eventually change the default to sort=False.

yes I think we should do this, but for all setops


> 2. Do a quick "break" from 0.24.0 to restore the 0.23.4 and earlier behavior with sort=False as the default. No deprecation cycle needed.

I don't think we should do this. what's done is done.

**Comment by TomAugspurger** [comment]
@jreback I think you’re the only one who prefers 1 (others may speak up). Are you OK with deferring to the majority on this one?

**Comment by jorisvandenbossche** [comment]
@jreback latest discussion/proposal is actually on the issue: https://github.com/pandas-dev/pandas/issues/24959#issuecomment-458264154

**Comment by jorisvandenbossche** [comment]
> > Implement a deprecation warning for the current behavior (sort=True), and eventually change the default to sort=False.
>
> yes I think we should do this, but for all setops

That is what Tom proposed in https://github.com/pandas-dev/pandas/issues/24959#issuecomment-458264154

**Comment by TomAugspurger** [comment]
To clarify 
https://github.com/pandas-dev/pandas/pull/24967#issuecomment-458552145, I think everyone is in favor of eventually getting to `sort=False` as the default, the question is how we get there, and most people seem to prefer `Index.intersection()` to have `sort=False` now, to match the behavior of pandas<0.24.0, and the future desired behavior.

**Comment by jreback** [comment]
I guess this is ok to push.

**Comment by jorisvandenbossche** [comment]
@TomAugspurger Thanks!

**Pull request merged**